File: 220831/b_1707_junh.py
import sys
from collections import deque


def check():
    colors = [None] * (V+1)

    def bfs(start):
        colors[start] = True
        q = deque([start])

        while q:
            a = q.popleft()
            clr = colors[a]

            for b in graph[a]:
                if colors[b] is None:
                    colors[b] = not clr
                    q.append(b)
                elif colors[b] == clr:
                    return False
                elif colors[b] != clr:
                    continue
        return True

    for i in range(1, V+1):
        if colors[i] is None:
            if not bfs(i):
                return "NO"

    return "YES"

input = sys.stdin.readline
K = int(input())
for _ in range(K):
    V, E = map(int, input().split())
    graph = [[] for _ in range(V+1)]

    for _ in range(E):
        a, b = map(int, input().split())
        graph[a].append(b)
        graph[b].append(a)

    print(check())

File: 220831/b_1707_junh.swift

for _ in 1...Int(readLine()!)! {
    let input = readLine()!.split(separator: " ").map {
        Int($0)!
    }
    let v = input[0]
    let e = input[1]
    print(solve(v: v, e: e))
}



func solve(v: Int, e: Int) -> String {
    var graph = Array(repeating: [Int](), count: v+1)
    
    for _ in 1...e {
        let nums = readLine()!.split(separator: " ").map {Int($0)!}
        graph[nums[0]].append(nums[1])
        graph[nums[1]].append(nums[0])
    }
    
    var colors = [Int](repeating: 0, count: v+1)
    
    func bfs(start: Int) -> Bool {
        colors[start] = 1
        
        var q = [start], idx = 0
        
        while idx < q.count {
            let n = q[idx]
            let clr = colors[n]
            
            for i in graph[n] {
                if colors[i] == clr {
                    return false
                } else if colors[i] == 0 {
                    colors[i] = -clr
                    q.append(i)
                }
            }
            idx += 1
        }
        return true
    }
    
    for i in 1...v {
        if colors[i] == 0 && bfs(start: i) == false{
            return "NO"
        }
    }
    return "YES"
}

